geoportal/geoportailv3_geoportal/views/wms.py

```diff
@@ -27,6 +31,64 @@ def _check_token(self, token):
             return True
         return False
 
+    def _get_arcgis_token(self, internal_wms, force_renew=False):
+        try:  # try to retrieve token from session storage
+            assert not force_renew
+            log.info('check token')
+            auth_token = self.request.session['auth_token']
+            log.info('token exists')
+            # renew token if it expires in the next 30s
+            assert datetime.fromtimestamp(float(auth_token['expires'])/1000 - 30) > datetime.now()
+            log.info('token still valid')
+            # systematic token check below is disabled
+            # token_check_data = urllib.parse.urlencode({'f': 'json', 'token': auth_token}).encode()
+            # token_check_request = urllib.request.Request(internal_wms.rest_url, data=token_check_data)
+            # try:
+            #     log.info(f"Check token at: {token_check_request.full_url}")
+            #     rep = urllib.request.urlopen(token_check_request, timeout=15)
+            #     check = json.load(rep)
+            #     if 'error' in check:
+            #         log.error(f"Token refused by: {token_check_request.full_url} - "
+            #                   f"server answered {check['error']}")
+            # except:
+            #     log.error(f"Failed token check at: {token_check_request.full_url}")
+            # assert 'error' not in check
+        except (KeyError, AssertionError) as e:
+            log.info('No valid token found in session storage')
+            config = self.request.registry.settings
+            token_data = {
+                'f': 'json',
+                'username': config["arcgis_token_username"],
+                'password': config["arcgis_token_password"],
+                'referer': 'x',
+                'expiration': 600
+            }
+            encoded_token_data = urllib.parse.urlencode(token_data).encode()
+            token_url = urllib.parse.urljoin(internal_wms.rest_url + '/', 'generateToken')
+            url_request = urllib.request.Request(token_url, data=encoded_token_data)
```

Comment:
Maybe using `requests` is more straight forward? (with `raise_on_status` and `json()`. Can also be mocked easily with https://pypi.org/project/responses/